### ISSUE TITLE
TFGrid monitoring (Phase 1)

### ISSUE THREAD

**Issue by despiegk** [threefoldtech/home#1294]
## Deliverables (internal)

- [x] Run Vector & node exporter on zos nodes @muhamadazmy  https://github.com/threefoldtech/zos/issues/1833
- [x] Run one or more vector node(s) to aggregate and potentially map data 
- [x] Run a big VM on the grid with Vector + Prometheus + Grafana 
~- [ ] Vector on Zos nodes scrapes local node exporter~
- [x] Vector on Zos then pushes to aggregator
- [x] Aggregator pushes to storage Vector
- [ ] Prometheus on storage nodes scrapes local Vector instance
- [x] Grafana with public readable node exporter dashboard connected to Prometheus

**Comment by xmonader** [comment]
Sabrina is off since today, story needs another owner

**Comment by delandtj** [comment]
monitoring of 3Nodes will involve things for ZOS too.
Monitoring is a really hairy subject.

I'm convinced that for the sheer size of what's to come we will need a layered system, and by all means push-based instead of pull based. 
Although it's a (little) bit more cumbersome to detect silent (died, unreachable, metrics daemon crash... ) hosts, it will be a lot easier to concatenate/mux/split streams of all sorts of metrics and info through a hierarchical stream processor (when size starts to matter) than having X pull based monitors scraping around through an unreliable network.

push-based over UDP through the standard internet instead of having to try to use Yggdrasil and have open metrics ports.

statsd comes to mind, together with Vector, where we can have many more observability streams aggregated than standard metrics provided by, say, Prometheus.

Prometheus has it's place, but not in a public grid.

**Comment by sasha-astiadi** [comment]
please determine which parts/increments that are feasible to accomplish by the end of v3.8.0 @rkhamis ...

**Comment by despiegk** [comment]
need an update here, what are the specs
who is doing it
link to tasks on other repo's

**Comment by sabrinasadik** [comment]
Looking into it with Bert, will update later today

**Comment by sabrinasadik** [comment]
After talking to Lee, Vector.dev might be a solution for this. 

Run on every zos node, which will collect info of all these nodes. 
Run a couple of vector nodes globally, zos nodes will send their info to these edge vector nodes. These will then aggregate and process this information before forwarding it to storage nodes. 

## Useful data to monitor

- [ ] HW specs usage: CPU, memory, disk space, disk I/O
- [ ] Current network throughput (bandwidth usage)
- [ ] Uptime 
- [ ] Current deployments + usage (x many VM's deployed on this node, x many networks,...)
- [ ] Node logs 

**Bandwidth and latency tests won't be that useful as it's dependent on momentary circumstances. Do we want to show this?** 

## Visualisation 

- [ ] Aggregated views over nodes, regions,.. 
**- Do we want everyone to be able to see what's running on your node?**
**- Ok to use Graffana for this?**

- [ ] User view
- Clicking on the node on the dashboard redirects to Grafana dashboard (this will take less development time if we don't implement it directly in the dashboard). 

## Data storage 

**- [ ] How long do we want to store this data? (data retention policy)** 
- [ ] On Grid: VM's

## Data pipeline 

- [ ] TBD after phase 1 since we don't have any metrics right now 

# Phase 1 
==> Run Vector & node exporter on zos nodes 
==> Run one or more vector node(s) to aggregate and potentially map data 
==> Run a big VM on the grid with Vector + Prometheus + Grafana 
==> Vector on Zos nodes scrapes local node exporter 
==> Vector on Zos then pushes to aggregator 
==> Aggregator pushes to storage Vector
==> Prometheus on storage nodes scrapes local Vector instance 
==> Grafana with public readable node exporter dashboard connected to Prometheus

**Comment by sabrinasadik** [comment]
Phase 1: https://github.com/threefoldtech/tf_operations/issues/1267

**Comment by LeeSmet** [comment]
We won't do this in phase 1, but we will probably do that in a future phase

**Comment by coesensbert** [comment]
Production environment deployed on the Grid: https://github.com/threefoldtech/tf_operations/issues/1307

Watch ZOS metrics at: https://metrics.grid.tf/
